File: API/1638_ARD_Order_Management/CS/ArdOrderManagementStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// Strategy based on DeMarker crossing a threshold.
/// </summary>
public class ArdOrderManagementStrategy : Strategy
{
	private readonly StrategyParam<int> _deMarkerPeriod;
	private readonly StrategyParam<decimal> _threshold;
	private readonly StrategyParam<DataType> _candleType;

	private DeMarker _deMarker;
	private decimal? _previousValue;

	/// <summary>
	/// DeMarker indicator period.
	/// </summary>
	public int DeMarkerPeriod
	{
		get => _deMarkerPeriod.Value;
		set => _deMarkerPeriod.Value = value;
	}

	/// <summary>
	/// Threshold for cross detection.
	/// </summary>
	public decimal Threshold
	{
		get => _threshold.Value;
		set => _threshold.Value = value;
	}

	/// <summary>
	/// Candle type to process.
	/// </summary>
	public DataType CandleType
	{
		get => _candleType.Value;
		set => _candleType.Value = value;
	}

	/// <summary>
	/// Initializes a new instance of <see cref="ArdOrderManagementStrategy"/>.
	/// </summary>
	public ArdOrderManagementStrategy()
	{
		_deMarkerPeriod = Param(nameof(DeMarkerPeriod), 2)
			.SetGreaterThanZero()
			.SetDisplay("DeMarker Period", "DeMarker indicator period", "Parameters")
			.SetCanOptimize(true);

		_threshold = Param(nameof(Threshold), 0.5m)
			.SetDisplay("Threshold", "DeMarker crossing level", "Parameters")
			.SetCanOptimize(true);

		_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(1).TimeFrame())
			.SetDisplay("Candle Type", "Type of candles", "General");
	}

	/// <inheritdoc />
	public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
	{
		return [(Security, CandleType)];
	}

	/// <inheritdoc />
	protected override void OnReseted()
	{
		base.OnReseted();

		_deMarker = default;
		_previousValue = default;
	}

	/// <inheritdoc />
	protected override void OnStarted(DateTimeOffset time)
	{
		base.OnStarted(time);

		StartProtection();

		_deMarker = new DeMarker { Length = DeMarkerPeriod };

		var subscription = SubscribeCandles(CandleType);
		subscription.Bind(_deMarker, ProcessCandle).Start();

		var area = CreateChartArea();
		if (area != null)
		{
			DrawCandles(area, subscription);
			DrawIndicator(area, _deMarker);
			DrawOwnTrades(area);
		}
	}

	private void ProcessCandle(ICandleMessage candle, decimal deMarkerValue)
	{
		if (candle.State != CandleStates.Finished)
			return;

		if (_previousValue is null)
		{
			_previousValue = deMarkerValue;
			return;
		}

		var buySignal = _previousValue > Threshold && deMarkerValue < Threshold;
		var sellSignal = _previousValue < Threshold && deMarkerValue > Threshold;

		if (buySignal && Position <= 0)
			BuyMarket();
		else if (sellSignal && Position >= 0)
			SellMarket();

		_previousValue = deMarkerValue;
	}
}

File: API/1638_ARD_Order_Management/PY/ard_order_management_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import DeMarker
from StockSharp.Algo.Strategies import Strategy


class ard_order_management_strategy(Strategy):
    def __init__(self):
        super(ard_order_management_strategy, self).__init__()
        self._de_marker_period = self.Param("DeMarkerPeriod", 14) \
            .SetDisplay("DeMarker Period", "DeMarker indicator period", "Parameters")
        self._threshold = self.Param("Threshold", 0.5) \
            .SetDisplay("Threshold", "DeMarker crossing level", "Parameters")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(4))) \
            .SetDisplay("Candle Type", "Type of candles", "General")
        self._previous_value = 0.0
        self._has_prev = False

    @property
    def de_marker_period(self):
        return self._de_marker_period.Value

    @property
    def threshold(self):
        return self._threshold.Value

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(ard_order_management_strategy, self).OnReseted()
        self._previous_value = 0.0
        self._has_prev = False

    def OnStarted(self, time):
        super(ard_order_management_strategy, self).OnStarted(time)
        de_marker = DeMarker()
        de_marker.Length = self.de_marker_period
        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(de_marker, self.on_process).Start()
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawIndicator(area, de_marker)
            self.DrawOwnTrades(area)

    def on_process(self, candle, de_marker_value):
        if candle.State != CandleStates.Finished:
            return
        if not self._has_prev:
            self._previous_value = de_marker_value
            self._has_prev = True
            return
        buy_signal = self._previous_value > self.threshold and de_marker_value < self.threshold
        sell_signal = self._previous_value < self.threshold and de_marker_value > self.threshold
        if buy_signal and self.Position <= 0:
            self.BuyMarket()
        elif sell_signal and self.Position >= 0:
            self.SellMarket()
        self._previous_value = de_marker_value

    def CreateClone(self):
        return ard_order_management_strategy()
